src/common/mfu_io.c

```diff
@@ -1380,7 +1395,63 @@ int mfu_truncate(const char* file, off_t length)
     return rc;
 }
 
-/* TODO: need to fix this function for dfs */
+int daos_truncate(const char* file, off_t length, mfu_file_t* mfu_file)
+{
+#ifdef DAOS_SUPPORT
+    char* name     = NULL;
+    char* dir_name = NULL;
+    parse_filename(file, &name, &dir_name);
+    assert(dir_name);
+
+    int rc = 0;
+
+    /* Lookup the parent directory */
+    dfs_obj_t* parent = daos_hash_lookup(dir_name, mfu_file);
+    if (parent == NULL) {
+        errno = ENOENT;
+        rc = -1;
+    } else {
+        /* open the obj in the parent */
+        dfs_obj_t* obj;
+        rc = dfs_open(mfu_file->dfs, parent, name,
+                      S_IFREG, O_RDWR,
+                      0, 0, NULL, &obj);
+        if (rc) {
+            MFU_LOG(MFU_LOG_ERR, "dfs_open %s failed (%d %s)",
+                    name, rc, strerror(rc));
+            errno = rc;
+            rc = -1;
+        } else {
+            /* truncate the obj */
+            daos_off_t offset = (daos_off_t) length;
+            rc = dfs_punch(mfu_file->dfs, obj, offset, DFS_MAX_FSIZE);
+            if (rc) {
+                MFU_LOG(MFU_LOG_ERR, "dfs_punch failed (%d %s)",
+                        rc, strerror(rc));
+                errno = rc;
+                rc = -1;
+            }
+
+            /* close the obj */
+            int tmp_rc = dfs_release(mfu_file->obj);
+            if (!rc && tmp_rc) {
+                MFU_LOG(MFU_LOG_ERR, "dfs_release %s failed (%d %s)",
+                        file, rc, strerror(rc));
+                errno = rc;
```

Comment:
Do you want to use tmp_rc here?

Reply:
Yes... It seems I will never learn my lesson with copy-pasting